File: ios/Services/Models/ProfileStudent.swift
//
//  StudentInfo.swift
//  ETSMobile WidgetExtension
//
//  Created by Jonathan Duval-Venne on 2022-11-07.
//

import Foundation
import SwiftyXMLParser

public let studentInfoPath = ["soap:Envelope", "soap:Body", "infoEtudiantResponse", "infoEtudiantResult"]
public let studentInfoErrorPath = ["soap:Envelope", "soap:Body", "infoEtudiantResponse", "infoEtudiantResult", "erreur"]

public class ProfileStudent {
    
    /// First name of the student
    let firstName: String?
    
    // Last name of the student
    let lastName: String?
    
    /// Permanent code of the student (XXXX00000000)
    let codePerm: String?
    
    /// Balance of the student
    let balance: String?
    
    public init(from xml:XML.Accessor) {
        self.firstName = xml.prenom.text
        self.lastName = xml.nom.text
        self.codePerm = xml.codePerm.text
        self.balance = xml.soldeTotal.text
    }
}

File: lib/features/app/signets-api/models/profile_student.dart
// FLUTTER / DART / THIRD-PARTIES
import 'package:xml/xml.dart';

class ProfileStudent {
  /// Balance of the student
  final String balance;

  /// First name of the student
  final String firstName;

  /// Last name of the student
  final String lastName;

  /// Permanent code of the student (XXXX00000000)
  final String permanentCode;

  ProfileStudent(
      {required this.balance,
      required this.firstName,
      required this.lastName,
      required this.permanentCode});

  /// Used to create a new [ProfileStudent] instance from a [XMLElement].
  factory ProfileStudent.fromXmlNode(XmlElement node) => ProfileStudent(
      lastName: node.getElement('nom')!.innerText.trimRight(),
      firstName: node.getElement('prenom')!.innerText.trimRight(),
      permanentCode: node.getElement('codePerm')!.innerText,
      balance: node.getElement('soldeTotal')!.innerText);

  /// Used to create [ProfileStudent] instance from a JSON file
  factory ProfileStudent.fromJson(Map<String, dynamic> map) => ProfileStudent(
      lastName: map['nom'] as String,
      firstName: map['prenom'] as String,
      permanentCode: map['codePerm'] as String,
      balance: map['soldeTotal'] as String);

  Map<String, dynamic> toJson() => {
        'nom': lastName.trimRight(),
        'prenom': firstName.trimRight(),
        'codePerm': permanentCode,
        'soldeTotal': balance
      };

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ProfileStudent &&
          runtimeType == other.runtimeType &&
          lastName == other.lastName &&
          firstName == other.firstName &&
          permanentCode == other.permanentCode &&
          balance == other.balance;

  @override
  int get hashCode =>
      lastName.hashCode ^
      firstName.hashCode ^
      permanentCode.hashCode ^
      balance.hashCode;
}
